core/orchestrator.go

```diff
@@ -234,12 +286,31 @@ func (n *LivepeerNode) transcodeAndCacheSeg(config transcodeConfig, ss *SignedSe
 	// Ensure length matches expectations. 4 second + 25% wiggle factor, 60fps
 	if err := ffmpeg.CheckMediaLen(fname, 4*1.25*1000, 60*4*1.25); err != nil {
 		glog.Errorf("Media length check failed: %v", err)
-		os.Remove(fname)
 		return terr(err)
 	}
+	seg.Name = fmt.Sprintf("%s_%d.ts", config.StrmID, seg.SeqNo)
+	url := fmt.Sprintf("%v/stream/%s", n.ServiceURI, seg.Name)
+	n.VideoSource.InsertHLSSegment(StreamID(config.StrmID), &seg) // XXX ensure we cleanup?
+
+	// Check if there's a transcoder available
+	var transcoder Transcoder
+	n.tcoderMutex.Lock()
+	if n.Transcoder == nil {
+		n.tcoderMutex.Unlock()
+		return terr(fmt.Errorf("No transcoders available on orchestrator"))
+	}
+	transcoder = n.Transcoder
+	n.tcoderMutex.Unlock()
+
+	// Small optimization so we aren't using http for local transcoding
+	// Replace by OS
+	if _, ok := transcoder.(*RemoteTranscoder); !ok {
+		url = fname
```

Comment:
For remote transcoders, we pass in a URL -- which is why we need to call `InsertHLSSegment` a few lines up. (This URL is downloaded directly by FFmpeg without an intermediate download + disk read.)

For local transcoders, we pass in the filename of the segment that was already downloaded.